File: STUDY/PYTHON/12_Proyecto_Integrador/ejercicios_extra.py
"""
Ejercicios extra - Proyecto integrador
"""
import json
from pathlib import Path


def guardar_tareas(path, tareas):
    """Guarda tareas en JSON."""
    # TODO: implementa la logica
    pass


def cargar_tareas(path):
    """Carga tareas o retorna lista vacia."""
    # TODO: implementa la logica
    return []


def borrar_tarea(tareas, task_id):
    """Elimina tarea por id y retorna True/False."""
    # TODO: implementa la logica
    return False


def next_id(tareas):
    """Retorna el siguiente id disponible."""
    # TODO: implementa la logica
    return 1


def buscar_tarea(tareas, task_id):
    """Retorna tarea por id o None."""
    # TODO: implementa la logica
    return None


def marcar_hecha(tareas, task_id):
    """Marca tarea como hecha y retorna True/False."""
    # TODO: implementa la logica
    return False


def filtrar_pendientes(tareas):
    """Retorna lista de tareas no hechas."""
    # TODO: implementa la logica
    return []


def resumen_tareas(tareas):
    """Retorna dict con conteos."""
    # TODO: implementa la logica
    return {}



def actualizar_titulo(tareas, task_id, nuevo):
    """Actualiza titulo de tarea y retorna True/False."""
    # TODO: implementa la logica
    return False


def ordenar_tareas(tareas, key="id"):
    """Retorna lista ordenada por key."""
    # TODO: implementa la logica
    return []

def main():
    base = Path(__file__).parent
    path = base / "tareas.json"
    tareas = [{"id": 1, "titulo": "leer", "hecha": False}]
    guardar_tareas(path, tareas)
    print("cargar:", cargar_tareas(path))
    print("borrar:", borrar_tarea(tareas, 1))


if __name__ == "__main__":
    main()

File: STUDY/RUST/11_Proyectos_Complejos/ejercicios_extra.rs
// ============================================
// EJERCICIOS EXTRA - PROYECTOS COMPLEJOS
// ============================================

#[derive(Debug, Clone)]
struct Tarea {
    id: u32,
    titulo: String,
    hecha: bool,
}

fn pendientes(tareas: &Vec<Tarea>) -> Vec<Tarea> {
    // TODO: filtra tareas no hechas
    Vec::new()
}

fn stats(tareas: &Vec<Tarea>) -> (usize, usize) {
    // TODO: retorna (hechas, pendientes)
    (0, 0)
}

fn buscar_por_id(tareas: &Vec<Tarea>, id: u32) -> Option<Tarea> {
    // TODO: retorna tarea clonada por id
    None
}

fn marcar_hecha(tareas: &mut Vec<Tarea>, id: u32) -> bool {
    // TODO: marca hecha y retorna True/False
    false
}

fn next_id(tareas: &Vec<Tarea>) -> u32 {
    // TODO: retorna siguiente id
    1
}

fn agregar_tarea(tareas: &mut Vec<Tarea>, titulo: &str) -> u32 {
    // TODO: agrega tarea con id nuevo y retorna id
    0
}

fn titulos_pendientes(tareas: &Vec<Tarea>) -> Vec<String> {
    // TODO: retorna titulos de pendientes
    Vec::new()
}


fn eliminar_tarea(tareas: &mut Vec<Tarea>, id: u32) -> bool {
    // TODO: elimina tarea por id
    false
}

fn resumen_texto(tareas: &Vec<Tarea>) -> String {
    // TODO: retorna resumen en texto
    String::new()
}

fn main() {
    let tareas = vec![
        Tarea { id: 1, titulo: "leer".to_string(), hecha: true },
        Tarea { id: 2, titulo: "escribir".to_string(), hecha: false },
    ];
    println!("pendientes: {:?}", pendientes(&tareas));
    println!("stats: {:?}", stats(&tareas));
}

